File: mekaneck/virtualbrain/operators/charge_ops.py
"""
Charge Operators: CONSERVE, REDISTRIBUTE.

Implements charge conservation and redistribution operations
for the charge-naming circuit isomorphism.

The soul (charge distribution) and consciousness (naming system)
are isomorphic structures operating on the same physical substrate.
"""

from typing import Tuple, Optional
import numpy as np


def CONSERVE(rho: np.ndarray) -> float:
    """
    CONSERVE operator: Compute total charge (verify conservation).

    integral(rho, V) = Q_total = const

    The total charge is conserved in the non-grounded circuit.

    Args:
        rho: Charge density array

    Returns:
        Total charge
    """
    return float(np.sum(rho))


def REDISTRIBUTE(
    rho: np.ndarray,
    target: np.ndarray,
    rate: float = 0.1,
) -> np.ndarray:
    """
    REDISTRIBUTE operator: Redistribute charge toward target.

    Moves charge to minimize variance while conserving total charge.

    Args:
        rho: Current charge distribution
        target: Target charge distribution
        rate: Redistribution rate

    Returns:
        Updated charge distribution (conserves total)
    """
    # Compute difference
    diff = target - rho

    # Apply rate-limited redistribution
    delta = rate * diff

    # Ensure conservation by removing mean
    delta = delta - np.mean(delta)

    return rho + delta


def charge_variance(rho: np.ndarray) -> float:
    """
    Compute charge variance.

    sigma^2 = <(rho - <rho>)^2>

    Lower variance indicates more uniform distribution.

    Args:
        rho: Charge density array

    Returns:
        Variance of charge distribution
    """
    return float(np.var(rho))


def minimize_variance(
    rho: np.ndarray,
    max_iterations: int = 100,
    tolerance: float = 1e-6,
) -> Tuple[np.ndarray, float]:
    """
    Redistribute charge to minimize variance.

    Variance minimization drives charge redistribution toward
    uniform distribution (equilibrium).

    Args:
        rho: Initial charge distribution
        max_iterations: Maximum iterations
        tolerance: Convergence tolerance

    Returns:
        (final_distribution, final_variance)
    """
    current = rho.copy()
    target = np.full_like(rho, np.mean(rho))

    for _ in range(max_iterations):
        var = charge_variance(current)
        if var < tolerance:
            break

        current = REDISTRIBUTE(current, target, rate=0.1)

    return current, charge_variance(current)


def charge_continuity(
    rho: np.ndarray,
    J: np.ndarray,
    dt: float,
    dx: float = 1.0,
) -> np.ndarray:
    """
    Apply charge continuity equation.

    d(rho)/dt + div(J) = 0

    Args:
        rho: Current charge density
        J: Current density (flux)
        dt: Time step
        dx: Spatial step

    Returns:
        Updated charge density
    """
    # Compute divergence of J (1D case)
    div_J = np.gradient(J, dx)

    # Update rho
    rho_new = rho - div_J * dt

    return rho_new


def couple_charge_consciousness(
    rho: np.ndarray,
    consciousness: float,
    coupling: float = 0.1,
) -> np.ndarray:
    """
    Couple charge distribution to consciousness level.

    The charge-naming isomorphism means consciousness modulates
    charge redistribution dynamics.

    Args:
        rho: Charge distribution
        consciousness: Consciousness level in [0, 1]
        coupling: Coupling strength

    Returns:
        Modulated charge distribution
    """
    # Higher consciousness -> faster equilibration
    rate = coupling * consciousness

    # Target is uniform distribution
    target = np.full_like(rho, np.mean(rho))

    return REDISTRIBUTE(rho, target, rate=rate)


def charge_equilibrium(
    rho: np.ndarray,
    temperature: float = 300.0,
) -> np.ndarray:
    """
    Compute thermal equilibrium charge distribution.

    Uses Boltzmann distribution for equilibrium.

    Args:
        rho: Current charge distribution
        temperature: Temperature in Kelvin

    Returns:
        Equilibrium distribution
    """
    from ..core.constants import BOLTZMANN_CONSTANT

    kT = BOLTZMANN_CONSTANT * temperature

    # Assume energy proportional to charge imbalance
    mean_rho = np.mean(rho)
    energy = (rho - mean_rho) ** 2

    # Boltzmann weights
    weights = np.exp(-energy / kT)
    weights = weights / np.sum(weights)

    # Equilibrium distribution
    total_charge = np.sum(rho)
    return weights * total_charge


def surface_coupling(
    rho_electric: np.ndarray,
    c_chemical: np.ndarray,
    coupling_constant: float = 1.0,
) -> float:
    """
    COUPLE_SC operator: Compute surface coupling potential.

    Couples electrical charge distribution to chemical concentration.

    Args:
        rho_electric: Electric charge density
        c_chemical: Chemical concentration
        coupling_constant: Coupling strength

    Returns:
        Surface coupling potential Phi
    """
    # Cross-correlation gives coupling strength
    Phi = coupling_constant * np.sum(rho_electric * c_chemical)
    return float(Phi)

File: virtual_brain/crates/vb-operators/src/charge_ops.rs
//! Charge Operators: CONSERVE, REDISTRIBUTE.
//!
//! Operators for charge transport and conservation.

use ndarray::Array1;
use vb_core::constants::BOLTZMANN_CONSTANT;

/// CONSERVE operator: Verify total charge conservation.
pub fn conserve(rho: &Array1<f64>) -> f64 {
    rho.sum()
}

/// REDISTRIBUTE operator: Move charge toward target distribution.
pub fn redistribute(rho: &Array1<f64>, target: &Array1<f64>, rate: f64) -> Array1<f64> {
    let diff = target - rho;
    rho + &(diff * rate)
}

/// Compute charge variance.
pub fn charge_variance(rho: &Array1<f64>) -> f64 {
    let mean = rho.mean().unwrap_or(0.0);
    rho.mapv(|x| (x - mean).powi(2)).mean().unwrap_or(0.0)
}

/// Minimize variance through redistribution.
pub fn minimize_variance(
    rho: &Array1<f64>,
    max_iterations: usize,
    tolerance: f64,
) -> Array1<f64> {
    let n = rho.len();
    let total = conserve(rho);
    let uniform = Array1::from_elem(n, total / n as f64);

    let mut current = rho.clone();

    for _ in 0..max_iterations {
        let var = charge_variance(&current);
        if var < tolerance {
            break;
        }
        current = redistribute(&current, &uniform, 0.1);
    }

    current
}

/// Charge continuity: drho/dt + div(J) = 0
pub fn charge_continuity(rho: &Array1<f64>, j: &Array1<f64>, dt: f64, dx: f64) -> Array1<f64> {
    let n = rho.len();
    let mut new_rho = rho.clone();

    for i in 1..n - 1 {
        let div_j = (j[i + 1] - j[i - 1]) / (2.0 * dx);
        new_rho[i] = rho[i] - div_j * dt;
    }

    // Boundary conditions (zero flux)
    new_rho[0] = new_rho[1];
    new_rho[n - 1] = new_rho[n - 2];

    new_rho
}

/// Couple charge to consciousness.
pub fn couple_charge_consciousness(
    rho: &Array1<f64>,
    consciousness: f64,
    coupling: f64,
) -> Array1<f64> {
    // Higher consciousness leads to more organized charge distribution
    let mean = rho.mean().unwrap_or(0.0);
    let factor = 1.0 - coupling * consciousness;

    rho.mapv(|x| mean + factor * (x - mean))
}

/// Boltzmann equilibrium distribution.
pub fn charge_equilibrium(rho: &Array1<f64>, temperature: f64) -> Array1<f64> {
    let n = rho.len();
    if n == 0 {
        return rho.clone();
    }

    // Simple harmonic potential centered at middle
    let center = n as f64 / 2.0;
    let potential: Array1<f64> = Array1::from_iter(
        (0..n).map(|i| 0.5 * (i as f64 - center).powi(2))
    );

    // Boltzmann factor
    let kt = BOLTZMANN_CONSTANT * temperature;
    let boltzmann: Array1<f64> = potential.mapv(|v| (-v / kt).exp());
    let z = boltzmann.sum();

    let total_charge = conserve(rho);
    boltzmann.mapv(|b| total_charge * b / z)
}

/// Surface coupling between electric and chemical.
pub fn surface_coupling(
    rho_electric: &Array1<f64>,
    c_chemical: &Array1<f64>,
    coupling_constant: f64,
) -> (Array1<f64>, Array1<f64>) {
    // Bidirectional coupling
    let interaction = rho_electric * c_chemical;

    let new_rho = rho_electric - &(&interaction * coupling_constant);
    let new_c = c_chemical + &(&interaction * coupling_constant);

    (new_rho, new_c)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_conserve() {
        let rho = Array1::from_vec(vec![1.0, 2.0, 3.0, 4.0]);
        assert!((conserve(&rho) - 10.0).abs() < 1e-10);
    }

    #[test]
    fn test_redistribute_conservation() {
        let rho = Array1::from_vec(vec![1.0, 2.0, 3.0, 4.0]);
        let target = Array1::from_vec(vec![2.5, 2.5, 2.5, 2.5]);

        let new_rho = redistribute(&rho, &target, 0.5);

        // Total charge should be conserved
        assert!((conserve(&new_rho) - conserve(&rho)).abs() < 1e-10);
    }

    #[test]
    fn test_minimize_variance() {
        let rho = Array1::from_vec(vec![1.0, 5.0, 2.0, 8.0]);
        let minimized = minimize_variance(&rho, 1000, 1e-6);

        // Variance should be reduced
        assert!(charge_variance(&minimized) < charge_variance(&rho));

        // Total charge conserved
        assert!((conserve(&minimized) - conserve(&rho)).abs() < 1e-10);
    }

    #[test]
    fn test_charge_variance() {
        let uniform = Array1::from_vec(vec![1.0, 1.0, 1.0, 1.0]);
        assert!(charge_variance(&uniform) < 1e-10);

        let varied = Array1::from_vec(vec![0.0, 2.0, 0.0, 2.0]);
        assert!(charge_variance(&varied) > 0.5);
    }
}
